power_grid_model_c/power_grid_model/include/power_grid_model/auxiliary/dataset.hpp

```diff
@@ -4,117 +4,263 @@
+    template <class StructType>
+    using DataStruct = std::conditional_t<is_data_mutable_v<dataset_type>, StructType, StructType const>;
 
-  public:
-    DataPointer() : ptr_{nullptr}, indptr_{nullptr}, batch_size_{}, elements_per_scenario_{} {}
-
-    // single batch dataset
-    DataPointer(ptr_t<void> ptr, Idx single_length)
-        : ptr_{ptr}, indptr_{nullptr}, batch_size_{1}, elements_per_scenario_{single_length} {}
-
-    // fix batch length
-    DataPointer(ptr_t<void> ptr, Idx batch_size, Idx elements_per_scenario)
-        : ptr_{ptr}, indptr_{nullptr}, batch_size_{batch_size}, elements_per_scenario_{elements_per_scenario} {}
-
-    // variable batches
-    DataPointer(ptr_t<void> ptr, Idx const* indptr, Idx batch_size)
-        : ptr_{ptr}, indptr_{indptr}, batch_size_{batch_size}, elements_per_scenario_{-1} {}
-
-    // copy to const constructor
-    DataPointer(ptr_t<void> ptr, Idx const* indptr, Idx batch_size, Idx elements_per_scenario)
-        : ptr_{ptr}, indptr_{indptr}, batch_size_{batch_size}, elements_per_scenario_{elements_per_scenario} {}
-
-    template <class T> std::pair<ptr_t<T>, ptr_t<T>> get_iterators(Idx pos) const {
-        assert(pos < batch_size_);
-        auto* const ptr = reinterpret_cast<ptr_t<T>>(ptr_);
-        if (indptr_) {
-            if (pos < 0) {
-                return std::make_pair(ptr, ptr + indptr_[batch_size_]);
-            }
-            return std::make_pair(ptr + indptr_[pos], ptr + indptr_[pos + 1]);
+    struct Buffer {
+        Data* data;
+        // for uniform buffer, indptr is empty
+        std::span<Indptr> indptr;
+    };
+
+    static constexpr Idx invalid_index{-1};
+
+    Dataset(bool is_batch, Idx batch_size, std::string_view dataset, MetaData const& meta_data)
+        : meta_data_{&meta_data},
+          dataset_info_{.is_batch = is_batch,
+                        .batch_size = batch_size,
+                        .dataset = &meta_data.get_dataset(dataset),
+                        .component_info = {}} {
+        if (!dataset_info_.is_batch && (dataset_info_.batch_size != 1)) {
+            throw DatasetError{"For non-batch dataset, batch size should be one!\n"};
         }
-        if (pos < 0) {
-            return std::make_pair(ptr, ptr + elements_per_scenario_ * batch_size_);
+    }
+
+    // implicit conversion constructor from writable to mutable, from writable and mutable to const
+    template <dataset_type_tag other_dataset_type>
+        requires((is_data_mutable_v<other_dataset_type> && !is_data_mutable_v<dataset_type>) ||
+                 (is_indptr_mutable_v<other_dataset_type> && !is_indptr_mutable_v<dataset_type>))
+    Dataset(Dataset<other_dataset_type> const& other)
+        : meta_data_{&other.meta_data()}, dataset_info_{other.get_description()} {
+        for (Idx i{}; i != other.n_components(); ++i) {
+            auto const& buffer = other.get_buffer(i);
+            buffers_.push_back(Buffer{.data = buffer.data, .indptr = buffer.indptr});
```

Comment:
It will not work because you are accessing `other.buffers_` which is a private member. Note the `other` has a differerent type. So the constructor of `this` type cannot access the private member of `other`.